### PR TITLE
Add array library.

### PR DESCRIPTION
**Pull request by KonovalovaVV**

### CODE REVIEW

File: array-library/arrayLibrary.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
+function ArrayHelper() {
+
+    function take(array, n) {
```
Comment (MaksHladki) [inline array-library/arrayLibrary.js:3]: Please, re-check this case: arrayHelper.take(testArray, -1))

File: array-library/arrayLibrary.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        skip: skip,
+        map: map,
+        reduce: reduce,
+        filter: filter,
+        foreach: foreach
+    }
+}
+
+ArrayChainLibrary = {
+    chain: function (array) {
```
Comment (MaksHladki) [inline array-library/arrayLibrary.js:38]: This method should be inside `ArrayHelper`.

Comment (MaksHladki) [reply array-library/arrayLibrary.js:38]: You can create a needed method inside the `ArrayHelper` function and call `ArrayChainLibrary` there.

File: array-library/arrayLibrary.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        return {
+            take: function (number) {
+                array = array.slice(0, number);
```
Comment (MaksHladki) [inline array-library/arrayLibrary.js:41]: Please, re-use logic from the `ArrayHelper` function. You've already declared there a set of static functions like `skip`, `take`, `map`, and so on.

File: array-library/arrayLibrary.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            filter: function (callback) {
+                array = array.filter(callback);
+                return this;
+            },
+            foreach: function (callback) {
+                array = array.forEach(callback);
+                return this;
+            },
+            value: function(){
+                return array;
```
Comment (MaksHladki) [inline array-library/arrayLibrary.js:65]: You don't have lazy evaluation. The main purpose of the task to calculate the result only when a user calls the `value` method. However, in your variant when the user calls this vacation they 've already had a result.

File: array-library/test.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
+var testArray = ["Item1", "Item2", "Item3"];
```
Comment (MaksHladki) [inline array-library/test.js:1]: please, rename the file to `arrayLibrary.test.es5.js`

File: array-library/test.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+const arrayHelper = new ArrayHelper();
+
+console.log("Take command result " + arrayHelper.take(testArray, 2));
+console.log("Skip command result " + arrayHelper.skip(testArray, 2));
+console.log("Map command result " + arrayHelper.map(testArray, mapFunction));
+console.log("Reduce command result " + arrayHelper.reduce(testArray, reduceFunction, "Sequence of elements: "));
+console.log("Filter command result " + arrayHelper.filter(testArray, filterFunction));
+console.log("ForEach command result " + arrayHelper.foreach(testArray, foreachFunction));
+
+console.log(ArrayChainLibrary.chain([1,2,3]).take(2).skip(1).value());
```
Comment (MaksHladki) [inline array-library/test.js:28]: Please, format the code. Each method from the chain should be on a separate line.

File: array-library/arrayLibrary.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
+function ArrayHelper() {
```
[the remark this answers is not in the document]
Comment (KonovalovaVV) [reply array-library/arrayLibrary.js:1]: Occasionally, removed.

[the remark this answers is not in the document]
Comment (KonovalovaVV) [reply array-library/arrayLibrary.js:1]: Ok, renamed the function to ArrayLibrary and the file to arrayLibrary.es5.js.

File: array-library/arrayLibrary.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+    function take(array, n) {
+        return array.slice(0, n);
+    }
+
+    function skip(array, n) {
```
[the remark this answers is not in the document]
Comment (KonovalovaVV) [reply array-library/arrayLibrary.js:7]: Return Item3 (the last element). Is it a wrong result?

[the remark this answers is not in the document]
Comment (KonovalovaVV) [reply array-library/arrayLibrary.js:7]: Fixed.

File: array-library/arrayLibrary.js (hunk 1 above)
Comment (KonovalovaVV) [reply array-library/arrayLibrary.js:3]: Fixed.

File: array-library/arrayLibrary.js (hunk 2 above)
Comment (KonovalovaVV) [reply array-library/arrayLibrary.js:38]: Ok.

File: array-library/arrayLibrary.js (hunk 3 above)
Comment (KonovalovaVV) [reply array-library/arrayLibrary.js:41]: Ok.

File: array-library/test.js (hunk 6 above)
Comment (KonovalovaVV) [reply array-library/test.js:28]: Ok.

File: array-library/test.js (hunk 5 above)
Comment (KonovalovaVV) [reply array-library/test.js:1]: Ok.

File: array-library/arrayLibrary.js (hunk 1 above)
Comment (MaksHladki) [reply array-library/arrayLibrary.js]: Your fix is wrong, you can use c# (linq) to check results.
arrayLibrary.take(testArray, -1) should return an empty array.

File: array-library/arrayLibrary.es5.js [hunk 9]
```diff
@@ -0,0 +1,111 @@
+function ArrayLibrary() {
+
+    function take(array, n) {
+        if (n < 0) {
```
Comment (MaksHladki) [inline array-library/arrayLibrary.es5.js:4]: Please, use a ternary operator instead of if/else.

File: array-library/arrayLibrary.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (MaksHladki) [reply array-library/arrayLibrary.js]: Please, double-check with c# linq your results.

File: array-library/arrayLibrary.es5.js [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    function foreach(array, callback) {
+        return array.forEach(callback);
+    }
+
+    function chain(array) {
+        var _queue = [];
+
+        return {
+            take: function (number) {
+                function fn(number) {
```
Comment (MaksHladki) [inline array-library/arrayLibrary.es5.js:40]: Try to use a shorter variant:
```
  take: function (number) {
                _queue.push(function() { return take(number)});

                return this;
            },
```

File: array-library/arrayLibrary.es5.js [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+                        array = skip(array, number);
+                    }
+                }
+                _queue.push(fn(number));
+                return this;
+            },
+            map: function (callback) {
+                function fn(callback) {
+                    return function () {
+                        array = map(array, callback);
```
Comment (MaksHladki) [inline array-library/arrayLibrary.es5.js:60]: I don't think it's a good idea to rewrite array every time.
```
 array = skip(array, number);
  array = map(array, callback);
  array = reduce(array, callback, initialValue);
```
Let's try to return a value and manipulate the array only in the`value` method.

File: array-library/arrayLibrary.es5.js [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+    function filter(array, callback) {
+        return array.filter(callback);
+    }
+
+    function foreach(array, callback) {
+        return array.forEach(callback);
+    }
+
+    function chain(array) {
```
Comment (MaksHladki) [inline array-library/arrayLibrary.es5.js:35]: Why did you remove the `ArrayChainLibrary` function? It was perfect.  I only would like to make it private.
You could call this like:
```
 function chain(array) {
        return new ArrayChainLibrary(array);
    }
```

Comment (KonovalovaVV) [reply array-library/arrayLibrary.es5.js:35]: Ok.

File: array-library/arrayLibrary.es5.js (hunk 10 above)
Comment (KonovalovaVV) [reply array-library/arrayLibrary.es5.js:40]: Ok.

File: array-library/arrayLibrary.es5.js (hunk 11 above)
Comment (KonovalovaVV) [reply array-library/arrayLibrary.es5.js:60]: Ok.

File: array-library/arrayLibrary.es5.js (hunk 9 above)
Comment (KonovalovaVV) [reply array-library/arrayLibrary.es5.js:4]: Ok.

File: array-library/arrayLibrary.es5.js [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+                foreach: function (callback) {
+                    _queue.push(function () {
+                        return foreach(array, callback)
+                    });
+
+                    return this;
+                },
+                value: function () {
+                    _queue.forEach(function (fn) {
+                        array = fn();
```
Comment (MaksHladki) [inline array-library/arrayLibrary.es5.js:79]: I guess we have to add an additional check for results from the `foreach` method.
Please, re-check
```
const arrayLibrary = new ArrayLibrary();
console.log(arrayLibrary
    .chain([1, 2, 3])
    .take(2)
    .map(mapFunction)
    .foreach(foreachFunction)
    .skip(1)
    .value());
```
